### PR TITLE
Basic collisions

### PR DESCRIPTION
**Pull request by Pathfinder-Liam**
PR for basic collision detection + addition of SpriteForFoo functionality that creates a two way reference between an object and it's associated sprite object

### CODE REVIEW

**Comment by cspotcode** [comment]
[Basic bullet-wall collision detection and response](https://trello.com/c/EvOVxZgQ/83-basic-bullet-wall-collision-detection-and-response)

File: src/weapon.py [hunk 1]
```diff
@@ -160,10 +158,29 @@ def update(
         self.time_since_shoot += delta_time
 
     def shoot(self, projectile_list: arcade.SpriteList):
-        bullet = arcade.SpriteSolidColor(10, 5, arcade.color.RED)
+        bullet: SpriteForMachineGun = SpriteForMachineGun(self)
```
Comment (not-mike-smith) [inline src/weapon.py:161]: why is a machine gun getting assigned to a variable named bullet?

File: src/bullet.py [hunk 2]
```diff
@@ -3,11 +3,21 @@
 from constants import SCREEN_HEIGHT, SCREEN_WIDTH
 from iron_math import add_vec2, rotate_vec2
 from player_manager import PlayerManager
+from typing import List, cast
+from arena.wall import SpriteForWall
 
 # For naming purposes, a bullet can be anything that comes out of a weapon included beams, rockets, etc
-def bullet_behavior(delta_time, player_manager: PlayerManager):
+def bullet_behavior(
```
Comment (not-mike-smith) [inline src/bullet.py:10]: Does this belong in a class?

File: src/weapon.py [hunk 3]
```diff
@@ -120,7 +118,7 @@ def update(
         self.time_since_shoot += delta_time
 
     def shoot(self, projectile_list: arcade.SpriteList):
-        rocket = arcade.SpriteSolidColor(30, 20, arcade.color.ORANGE)
+        rocket: SpriteForRocket = SpriteForRocket(self)
```
Comment (not-mike-smith) [inline src/weapon.py:121]: 👍

File: src/bullet.py [hunk 4]
```diff
@@ -3,11 +3,21 @@
 from constants import SCREEN_HEIGHT, SCREEN_WIDTH
 from iron_math import add_vec2, rotate_vec2
 from player_manager import PlayerManager
+from typing import List, cast
+from arena.wall import SpriteForWall
 
 # For naming purposes, a bullet can be anything that comes out of a weapon included beams, rockets, etc
```
Comment (not-mike-smith) [inline src/bullet.py:9]: Is this comment still true?  I think we discussed updating the name "bullet" here in a previous meeting, but I don't know the card that had the output of the discussion.

File: src/bullet.py [hunk 5]
```diff
@@ -3,11 +3,21 @@
 from constants import SCREEN_HEIGHT, SCREEN_WIDTH
 from iron_math import add_vec2, rotate_vec2
 from player_manager import PlayerManager
+from typing import List, cast
+from arena.wall import SpriteForWall
```
Comment (not-mike-smith) [inline src/bullet.py:7]: Why does this import SpriteForWall instead of Wall?

**Review by not-mike-smith — APPROVED** [review]
Approved b/c I have questions, but don't see anything wrong.  Happy to rubber-stamp further changes if my questions lead to updates.

As Andrew mentioned in passing in the meeting, I think there's a missing class that describes a game element.  An object that exists in the game which has at least one sprite associated with it, and can interact with other elements.  Cars with walls, walls with bullets, etc.  These objects exist in virtual space and should have some calculable inertia.  While the sprite is being used to check for whether an interaction occurs, the behavior of the resulting interaction seems conceptually different from the sprite.  I think this is something for another ticket, but relevant to my questions about what class is being imported.

File: src/weapon.py (hunk 1 above)
Comment (cspotcode) [reply src/weapon.py:161]: Probably `SpriteForMachineGun` is actually a sprite for a machine gun bullet, nor for the gun itself.  Maybe we tweak the naming to make this clearer?

Comment (Thorakii) [reply src/weapon.py:161]: I believe the next step will be making classes for these projectiles and beams, so these will all morph into either `SpriteForProjectile` and `SpriteForBeam`, so the naming convention issue will go away

**Review by Thorakii — APPROVED** [review]
This looks great to me

File: src/bullet.py (hunk 4 above)
Comment (Thorakii) [reply src/bullet.py:9]: I added this comment. I had started using the convention `weapon_output` for anything that comes out of a weapon. It felt really word and also confusing as `weapon_output_behavior` sounds like it's going to output a behavior. just using the term bullet to encompass projectiles, beams, mines etc is much cleaner if less accurate. The comment is meant to make this true for us.

**Pull request merged**